test/correctness.cpp

```diff
@@ -0,0 +1,422 @@
+
+#include "../src/nodetypeinfo.h"
+#include "../src/correctness.h"
+
+using namespace piwcs::prw;
+
+using SType = Section::AllowedTravel;
+
+struct SectionTestInfo{
+    Section::AllowedTravel type;
+    bool inward = false; // Section is directed into node if `true`
+};
+
+template<std::size_t N>
+bool localCorrectnessMinimal(
+        NodeType type,
+        const SectionTestInfo (& sections)[N]
+        ){
+
+    assert(
+            N>=(type->slotCount) &&
+            "Invalid section array provided, slot counts don't match!"
+            );
+
+    Model model;
+    model.newNode(type, "n0");
+
+    for(SlotId slot=0; slot<(type->slotCount); slot++){
+        Identifier nodeId = std::format("n{}", slot+1);
+        Identifier sectionId = std::format("s{}", slot);
+        model.newNode(type, nodeId);
+        model.newSection(sectionId, sections[slot].type, nullptr);
+
+        if(sections[slot].inward){
+            model.link(sectionId, nodeId, 0, "n0", slot);
+        }
+        else{
+            model.link(sectionId, "n0", slot, nodeId, 0);
+        }
+    }
+
+    return isLocallyCorrect(model, "n0");
+}
+
+#define UNWRAP(...) __VA_ARGS__
+#define INCORRECT_TEST(testName, nodeType, sectionInfo)             \
+    TEST(LocalCorrectness, nodeType ##_## testName ## _IsIncorrect){\
+        bool res = localCorrectnessMinimal(                         \
+                nodeType,                                           \
+                UNWRAP sectionInfo                                  \
+            );                                                      \
+   ASSERT_EQ(res, false);                                           \
+    }                                                               \
+
+#define CORRECT_TEST(testName, nodeType, sectionInfo)               \
+    TEST(LocalCorrectness, nodeType ##_## testName ## _IsCorrect){  \
+        bool res = localCorrectnessMinimal(                         \
+                nodeType,                                           \
+                UNWRAP sectionInfo                                  \
+            );                                                      \
+        ASSERT_EQ(res, true);                                       \
+    }                                                               \
+
+// << "Expected correct node of type `" << #nodeType << "`";
+
+std::vector<NodeType> types = {
+        END,
+        THRU,
+        CROSSING,
+        MOTORIZED,
+        FIXED,
+        PASSIVE,
+        MANUAL
+};
+
+TEST(LocalCorrectness, AnyCompletelyForbidden_IsCorrect){
+
+    SectionTestInfo sections[Node::MAX_SLOTS];
+
+    for (SlotId slot=0; slot<Node::MAX_SLOTS; ++slot){
+        sections[slot] = {SType::NONE};
+    }
+
+    for(NodeType type : types){
+
+        bool res = localCorrectnessMinimal(
+                type,
+                sections
+        );
+        ASSERT_TRUE(res) << "Expected correct node of type `" << type->name << "`";
+    }
+}
+
```

Comment:
Test for correct error handling in `isLocallyCorrect`: non-existing ID, `ID_NULL`, empty Model, incomplete Model

Reply:
Added tests for invalid ids and empty Model in 37e1a64336c8d284350b89fce390041d773c2a00. Test for incomplete Model (a node with empty slot) was added in 574bd5a39fc3273895762c5b3d6b2128937c9202 ([code](https://github.com/Lipovr/PiwcsPrwModel/blame/37e1a64336c8d284350b89fce390041d773c2a00/test/correctness.cpp#L67-L71C2)).